File: sdk/typescript/src/workspace.ts
import type { GitsliceClient } from "./client.js";
import {
  parseDiffResult,
  parseMergeResult,
  parseSearchMatch,
  parseSnapshotInfo,
  parseWorkspaceConflict,
  parseWorkspaceEntry,
  type DiffResult,
  type MergeResult,
  type SearchMatch,
  type SnapshotInfo,
  type WorkspaceConflict,
  type WorkspaceEntry,
  type WriteResult,
} from "./types.js";

type ApiRecord = Record<string, unknown>;

export interface MergeOptions {
  sourceSnapshotId?: string;
  targetSnapshotId?: string;
  baseWorkspaceId?: string;
  baseSnapshotId?: string;
  message?: string;
}

export interface ForkOptions {
  name?: string;
  description?: string;
  snapshotId?: string;
}

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

interface BufferLike {
  toString(encoding: "base64"): string;
}

interface BufferConstructorLike {
  from(data: Uint8Array): BufferLike;
  from(data: string, encoding: "base64"): Uint8Array;
}

export class Workspace {
  readonly workspaceId: string;

  constructor(private readonly client: GitsliceClient, workspaceId: string) {
    this.workspaceId = workspaceId;
  }

  async read(path: string): Promise<string> {
    return textDecoder.decode(await this.readBytes(path));
  }

  async readBytes(path: string): Promise<Uint8Array> {
    const payload = await this.client.requestJson("GET", this.filePath(path));
    return decodeBytes(String(payload.content ?? ""));
  }

  async write(path: string, content: string | Uint8Array): Promise<WriteResult> {
    const payload = await this.client.requestJson("PUT", this.filePath(path), {
      workspaceId: this.workspaceId,
      path,
      content: encodeBytes(content),
    });
    return {
      workspaceId: String(payload.workspaceId ?? ""),
      path: String(payload.path ?? ""),
      size: Number(payload.size ?? 0),
      hash: String(payload.hash ?? ""),
      commitHash: String(payload.commitHash ?? ""),
    };
  }

  async delete(): Promise<ApiRecord> {
    return this.client.deleteWorkspace(this.workspaceId);
  }

  async rm(path: string): Promise<ApiRecord> {
    return this.client.requestJson("DELETE", this.filePath(path));
  }

  async mkdir(path: string): Promise<ApiRecord> {
    const escapedPath = encodePath(path);
    return this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/mkdir/${escapedPath}`,
      { workspaceId: this.workspaceId, path },
    );
  }

  async ls(path = ""): Promise<WorkspaceEntry[]> {
    let requestPath = `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/ls`;
    if (path) {
      requestPath = `${requestPath}/${encodePath(path)}`;
    }
    const payload = await this.client.requestJson("GET", requestPath);
    const entries = Array.isArray(payload.entries) ? payload.entries : [];
    return entries.map((item) => parseWorkspaceEntry((item as ApiRecord) ?? {}));
  }

  async exists(path: string): Promise<boolean> {
    const payload = await this.client.requestJson("GET", this.pathRoute("exists", path));
    return Boolean(payload.exists ?? false);
  }

  async stat(path: string): Promise<WorkspaceEntry | null> {
    const payload = await this.client.requestJson("GET", this.pathRoute("stat", path));
    if (!payload.exists) {
      return null;
    }
    return parseWorkspaceEntry((payload.entry as ApiRecord | undefined) ?? {});
  }

  async mv(sourcePath: string, destinationPath: string): Promise<ApiRecord> {
    return this.client.requestJson("POST", `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:move`, {
      workspaceId: this.workspaceId,
      sourcePath,
      destinationPath,
    });
  }

  async cp(sourcePath: string, destinationPath: string): Promise<ApiRecord> {
    return this.client.requestJson("POST", `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:copy`, {
      workspaceId: this.workspaceId,
      sourcePath,
      destinationPath,
    });
  }

  async readMany(paths: string[]): Promise<Record<string, string | null>> {
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:readFiles`,
      { workspaceId: this.workspaceId, paths },
    );
    const files = Array.isArray(payload.files) ? payload.files : [];
    const result: Record<string, string | null> = {};
    for (const file of files) {
      const item = (file as ApiRecord) ?? {};
      const pathKey = String(item.path ?? "");
      result[pathKey] = item.found ? textDecoder.decode(decodeBytes(String(item.content ?? ""))) : null;
    }
    return result;
  }

  async writeMany(files: Record<string, string | Uint8Array>): Promise<WriteResult[]> {
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:writeFiles`,
      {
        workspaceId: this.workspaceId,
        files: Object.entries(files).map(([path, content]) => ({
          path,
          content: encodeBytes(content),
        })),
      },
    );
    const results = Array.isArray(payload.files) ? payload.files : [];
    const commitHash = String(payload.commitHash ?? "");
    return results.map((file) => {
      const item = (file as ApiRecord) ?? {};
      return {
        workspaceId: this.workspaceId,
        path: String(item.path ?? ""),
        size: Number(item.size ?? 0),
        hash: String(item.hash ?? ""),
        commitHash,
      };
    });
  }

  async glob(pattern: string): Promise<string[]> {
    const payload = await this.client.requestJson(
      "GET",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:glob`,
      undefined,
      { pattern },
    );
    return Array.isArray(payload.paths) ? payload.paths.map((item) => String(item)) : [];
  }

  async search(query: string, glob?: string): Promise<SearchMatch[]> {
    const payload = await this.client.requestJson(
      "GET",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:search`,
      undefined,
      { query, glob },
    );
    const matches = Array.isArray(payload.matches) ? payload.matches : [];
    return matches.map((item) => parseSearchMatch((item as ApiRecord) ?? {}));
  }

  async snapshot(message: string): Promise<SnapshotInfo> {
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/snapshot`,
      { workspaceId: this.workspaceId, message },
    );
    return parseSnapshotInfo((payload.snapshot as ApiRecord | undefined) ?? {});
  }

  async snapshots(limit?: number): Promise<SnapshotInfo[]> {
    const payload = await this.client.requestJson(
      "GET",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/snapshots`,
      undefined,
      { limit },
    );
    const snapshots = Array.isArray(payload.snapshots) ? payload.snapshots : [];
    return snapshots.map((item) => parseSnapshotInfo((item as ApiRecord) ?? {}));
  }

  async restore(snapshotId: string, message = ""): Promise<SnapshotInfo> {
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/restore/${encodeURIComponent(snapshotId)}`,
      {
        workspaceId: this.workspaceId,
        snapshotId,
        message,
      },
    );
    return parseSnapshotInfo((payload.snapshot as ApiRecord | undefined) ?? {});
  }

  async diff(fromSnapshotId: string, toSnapshotId?: string, includePatches = true): Promise<DiffResult> {
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}:diff`,
      {
        workspaceId: this.workspaceId,
        fromSnapshotId,
        toSnapshotId: toSnapshotId ?? "",
        includePatches,
      },
    );
    return parseDiffResult(payload);
  }

  async fork(forkWorkspaceId: string, options: ForkOptions = {}): Promise<Workspace> {
    await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/fork`,
      {
        workspaceId: this.workspaceId,
        forkWorkspaceId,
        name: options.name ?? forkWorkspaceId,
        description: options.description ?? "",
        snapshotId: options.snapshotId ?? "",
      },
    );
    return new Workspace(this.client, forkWorkspaceId);
  }

  async merge(sourceWorkspace: string | Workspace, options: MergeOptions = {}): Promise<MergeResult> {
    const sourceWorkspaceId =
      typeof sourceWorkspace === "string" ? sourceWorkspace : sourceWorkspace.workspaceId;
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/merge`,
      {
        workspaceId: this.workspaceId,
        sourceWorkspaceId,
        sourceSnapshotId: options.sourceSnapshotId ?? "",
        targetSnapshotId: options.targetSnapshotId ?? "",
        baseWorkspaceId: options.baseWorkspaceId ?? "",
        baseSnapshotId: options.baseSnapshotId ?? "",
        message: options.message ?? "",
      },
    );
    return parseMergeResult(payload);
  }

  async conflicts(otherWorkspaceId?: string): Promise<WorkspaceConflict[]> {
    const payload = await this.client.requestJson(
      "GET",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/conflicts`,
      undefined,
      { otherWorkspaceId },
    );
    const conflicts = Array.isArray(payload.conflicts) ? payload.conflicts : [];
    return conflicts.map((item) => parseWorkspaceConflict((item as ApiRecord) ?? {}));
  }

  async resolveConflict(path: string, preferredWorkspace?: string | Workspace): Promise<WorkspaceConflict> {
    const preferredWorkspaceId =
      typeof preferredWorkspace === "string"
        ? preferredWorkspace
        : preferredWorkspace?.workspaceId ?? this.workspaceId;
    const payload = await this.client.requestJson(
      "POST",
      `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/conflicts:resolve`,
      {
        workspaceId: this.workspaceId,
        path,
        preferredWorkspaceId,
      },
    );
    return parseWorkspaceConflict((payload.conflict as ApiRecord | undefined) ?? {});
  }

  private filePath(path: string): string {
    return `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/files/${encodePath(path)}`;
  }

  private pathRoute(route: string, path: string): string {
    const base = `/v1/fs/workspaces/${encodeURIComponent(this.workspaceId)}/${route}`;
    return path ? `${base}/${encodePath(path)}` : base;
  }
}

function encodePath(path: string): string {
  return path
    .split("/")
    .filter((segment) => segment.length > 0)
    .map((segment) => encodeURIComponent(segment))
    .join("/");
}

function encodeBytes(content: string | Uint8Array): string {
  const bytes = typeof content === "string" ? textEncoder.encode(content) : content;
  const nodeBuffer = (globalThis as typeof globalThis & { Buffer?: BufferConstructorLike }).Buffer;
  if (nodeBuffer) {
    return nodeBuffer.from(bytes).toString("base64");
  }
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

function decodeBytes(content: string): Uint8Array {
  const nodeBuffer = (globalThis as typeof globalThis & { Buffer?: BufferConstructorLike }).Buffer;
  if (nodeBuffer) {
    return new Uint8Array(nodeBuffer.from(content, "base64"));
  }
  const binary = atob(content);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes;
}

File: sdk/python/gitslice/workspace.py
from __future__ import annotations

import base64
from typing import Any, Optional
from urllib import parse

from .types import (
    DiffResult,
    MergeResult,
    SearchMatch,
    SnapshotInfo,
    WorkspaceConflict,
    WorkspaceEntry,
    WriteResult,
)


class Workspace:
    def __init__(self, client: Any, workspace_id: str, *, cleanup_on_exit: bool = False) -> None:
        self._client = client
        self.workspace_id = workspace_id
        self._cleanup_on_exit = cleanup_on_exit

    def __enter__(self) -> "Workspace":
        return self

    def __exit__(self, exc_type: Any, exc: Any, tb: Any) -> bool:
        if not self._cleanup_on_exit:
            return False
        try:
            self.delete()
        except Exception:
            if exc is None:
                raise
        return False

    def read(self, path: str, *, encoding: str = "utf-8") -> str:
        return self.read_bytes(path).decode(encoding)

    def read_bytes(self, path: str) -> bytes:
        payload = self._client.request_json("GET", self._file_path(path))
        return self._decode_bytes(payload["content"])

    def write(self, path: str, content: str | bytes, *, encoding: str = "utf-8") -> WriteResult:
        payload = self._client.request_json(
            "PUT",
            self._file_path(path),
            payload={
                "workspaceId": self.workspace_id,
                "path": path,
                "content": self._encode_bytes(content, encoding=encoding),
            },
        )
        return WriteResult.from_api(payload)

    def delete(self) -> dict[str, Any]:
        return self._client.delete_workspace(self.workspace_id)

    def rm(self, path: str) -> dict[str, Any]:
        return self._client.request_json("DELETE", self._file_path(path))

    def mkdir(self, path: str) -> dict[str, Any]:
        escaped = parse.quote(path.strip("/"), safe="/")
        return self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/mkdir/{escaped}",
            payload={"workspaceId": self.workspace_id, "path": path},
        )

    def ls(self, path: str = "") -> list[WorkspaceEntry]:
        request_path = f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/ls"
        if path:
            request_path += f"/{parse.quote(path.strip('/'), safe='/')}"
        payload = self._client.request_json("GET", request_path)
        return [WorkspaceEntry.from_api(item) for item in payload.get("entries", [])]

    def exists(self, path: str) -> bool:
        payload = self._client.request_json("GET", self._path_route("exists", path))
        return bool(payload.get("exists", False))

    def stat(self, path: str) -> Optional[WorkspaceEntry]:
        payload = self._client.request_json("GET", self._path_route("stat", path))
        if not payload.get("exists", False):
            return None
        return WorkspaceEntry.from_api(payload.get("entry", {}))

    def mv(self, source_path: str, destination_path: str) -> dict[str, Any]:
        return self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:move",
            payload={
                "workspaceId": self.workspace_id,
                "sourcePath": source_path,
                "destinationPath": destination_path,
            },
        )

    def cp(self, source_path: str, destination_path: str) -> dict[str, Any]:
        return self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:copy",
            payload={
                "workspaceId": self.workspace_id,
                "sourcePath": source_path,
                "destinationPath": destination_path,
            },
        )

    def read_many(self, paths: list[str], *, encoding: str = "utf-8") -> dict[str, Optional[str]]:
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:readFiles",
            payload={"workspaceId": self.workspace_id, "paths": paths},
        )
        result: dict[str, Optional[str]] = {}
        for item in payload.get("files", []):
            if item.get("found"):
                result[str(item.get("path", ""))] = self._decode_bytes(item.get("content", "")).decode(encoding)
            else:
                result[str(item.get("path", ""))] = None
        return result

    def write_many(self, files: dict[str, str | bytes], *, encoding: str = "utf-8") -> list[WriteResult]:
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:writeFiles",
            payload={
                "workspaceId": self.workspace_id,
                "files": [
                    {"path": path, "content": self._encode_bytes(content, encoding=encoding)}
                    for path, content in files.items()
                ],
            },
        )
        commit_hash = str(payload.get("commitHash", ""))
        results = []
        for item in payload.get("files", []):
            results.append(
                WriteResult(
                    workspace_id=self.workspace_id,
                    path=str(item.get("path", "")),
                    size=int(item.get("size", 0)),
                    hash=str(item.get("hash", "")),
                    commit_hash=commit_hash,
                )
            )
        return results

    def glob(self, pattern: str) -> list[str]:
        payload = self._client.request_json(
            "GET",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:glob",
            params={"pattern": pattern},
        )
        return [str(item) for item in payload.get("paths", [])]

    def search(self, query: str, *, glob: Optional[str] = None) -> list[SearchMatch]:
        params = {"query": query}
        if glob:
            params["glob"] = glob
        payload = self._client.request_json(
            "GET",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:search",
            params=params,
        )
        return [SearchMatch.from_api(item) for item in payload.get("matches", [])]

    def snapshot(self, message: str) -> SnapshotInfo:
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/snapshot",
            payload={"workspaceId": self.workspace_id, "message": message},
        )
        return SnapshotInfo.from_api(payload.get("snapshot", {}))

    def snapshots(self, *, limit: Optional[int] = None) -> list[SnapshotInfo]:
        payload = self._client.request_json(
            "GET",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/snapshots",
            params={"limit": limit},
        )
        return [SnapshotInfo.from_api(item) for item in payload.get("snapshots", [])]

    def restore(self, snapshot_id: str, *, message: str = "") -> SnapshotInfo:
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/restore/{parse.quote(snapshot_id, safe='')}",
            payload={
                "workspaceId": self.workspace_id,
                "snapshotId": snapshot_id,
                "message": message,
            },
        )
        return SnapshotInfo.from_api(payload.get("snapshot", {}))

    def diff(self, from_snapshot_id: str, *, to_snapshot_id: Optional[str] = None, include_patches: bool = True) -> DiffResult:
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}:diff",
            payload={
                "workspaceId": self.workspace_id,
                "fromSnapshotId": from_snapshot_id,
                "toSnapshotId": to_snapshot_id or "",
                "includePatches": include_patches,
            },
        )
        return DiffResult.from_api(payload)

    def fork(
        self,
        fork_workspace_id: str,
        *,
        name: Optional[str] = None,
        description: str = "",
        snapshot_id: Optional[str] = None,
        cleanup_on_exit: bool = False,
    ) -> "Workspace":
        self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/fork",
            payload={
                "workspaceId": self.workspace_id,
                "forkWorkspaceId": fork_workspace_id,
                "name": name or fork_workspace_id,
                "description": description,
                "snapshotId": snapshot_id or "",
            },
        )
        return Workspace(self._client, fork_workspace_id, cleanup_on_exit=cleanup_on_exit)

    def merge(
        self,
        source_workspace: str | "Workspace",
        *,
        source_snapshot_id: Optional[str] = None,
        target_snapshot_id: Optional[str] = None,
        base_workspace_id: Optional[str] = None,
        base_snapshot_id: Optional[str] = None,
        message: str = "",
    ) -> MergeResult:
        source_workspace_id = (
            source_workspace.workspace_id if isinstance(source_workspace, Workspace) else source_workspace
        )
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/merge",
            payload={
                "workspaceId": self.workspace_id,
                "sourceWorkspaceId": source_workspace_id,
                "sourceSnapshotId": source_snapshot_id or "",
                "targetSnapshotId": target_snapshot_id or "",
                "baseWorkspaceId": base_workspace_id or "",
                "baseSnapshotId": base_snapshot_id or "",
                "message": message,
            },
        )
        return MergeResult.from_api(payload)

    def conflicts(self, *, other_workspace_id: Optional[str] = None) -> list[WorkspaceConflict]:
        params = {"otherWorkspaceId": other_workspace_id} if other_workspace_id else None
        payload = self._client.request_json(
            "GET",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/conflicts",
            params=params,
        )
        return [WorkspaceConflict.from_api(item) for item in payload.get("conflicts", [])]

    def resolve_conflict(
        self,
        path: str,
        *,
        preferred_workspace: Optional[str | "Workspace"] = None,
    ) -> WorkspaceConflict:
        preferred_workspace_id = (
            preferred_workspace.workspace_id
            if isinstance(preferred_workspace, Workspace)
            else preferred_workspace
        )
        payload = self._client.request_json(
            "POST",
            f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/conflicts:resolve",
            payload={
                "workspaceId": self.workspace_id,
                "path": path,
                "preferredWorkspaceId": preferred_workspace_id or self.workspace_id,
            },
        )
        return WorkspaceConflict.from_api(payload.get("conflict", {}))

    def _file_path(self, path: str) -> str:
        return f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/files/{parse.quote(path.strip('/'), safe='/')}"

    def _path_route(self, route: str, path: str) -> str:
        base = f"/v1/fs/workspaces/{parse.quote(self.workspace_id, safe='')}/{route}"
        if not path:
            return base
        return f"{base}/{parse.quote(path.strip('/'), safe='/')}"

    @staticmethod
    def _encode_bytes(content: str | bytes, *, encoding: str) -> str:
        data = content.encode(encoding) if isinstance(content, str) else content
        return base64.b64encode(data).decode("ascii")

    @staticmethod
    def _decode_bytes(content: str) -> bytes:
        return base64.b64decode(content)
